oak_client/src/attestation.rs

```diff
@@ -0,0 +1,156 @@
+}
+
+/// Custom X.509 certificate verifier that checks TEE quotes in the X.509 TEE extension.
+struct RemoteAttestationVerifier {
+    tee_quote: String,
+}
+
+impl RemoteAttestationVerifier {
+    fn new(tee_quote: &str) -> Self {
+        Self {
+            tee_quote: tee_quote.to_string(),
+        }
+    }
+
+    fn check_tee_quote(&self, extension: &str) -> anyhow::Result<()> {
+        let pattern = r"^Quote:(.*)$";
+        let re = regex::Regex::new(&pattern).expect("Incorrect regular expression");
+        let tee_quotes = re.captures(&extension).context("Couldn't find TEE quote")?;
+
+        // `0` capture group corresponds to the whole pattern, while `1` corresponds to the first
+        // match.
+        // https://docs.rs/regex/1.4.3/regex/struct.Captures.html#method.get
+        if self.tee_quote == tee_quotes.get(1).map_or("", |m| m.as_str()) {
+            Ok(())
+        } else {
+            Err(anyhow!("Incorrect TEE quote"))
+        }
+    }
+}
+
+impl rustls::ServerCertVerifier for RemoteAttestationVerifier {
+    fn verify_server_cert(
+        &self,
+        roots: &rustls::RootCertStore,
+        certs: &[rustls::Certificate],
+        hostname: webpki::DNSNameRef,
+        ocsp: &[u8],
+    ) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
+        if certs.len() != 1 {
+            return Err(rustls::TLSError::General(
+                "Only a single certificate should be provided".to_string(),
+            ));
+        }
+        let certificate = &certs[0].0;
+
+        // Parse X.509 certificate encoded using DER format.
+        // https://tools.ietf.org/html/rfc7468
+        let (_, parsed_certificate) = parse_x509_certificate(&certificate).map_err(|error| {
+            rustls::TLSError::General(format!("Couldn't parse certificate: {}", error))
+        })?;
+
+        // Extract X.509 TEE extension.
+        let extensions = parsed_certificate.extensions();
+        let tee_extension = extensions.get(&TEE_EXTENSION_OID).ok_or_else(|| {
+            rustls::TLSError::General("Couldn't find TEE quote X.509 extension".to_string())
+        })?;
+        let tee_extension_value = parse_der(tee_extension.value)
+            .map_err(|error| {
+                rustls::TLSError::General(format!("Couldn't parse TEE extension: {:?}", error))
+            })?
+            .1
+            .as_str()
+            .map_err(|error| {
+                rustls::TLSError::General(format!("Couldn't parse TEE quote value: {:?}", error))
+            })?;
+
+        // Check that X.509 TEE extension contains a correct TEE quote.
+        self.check_tee_quote(tee_extension_value).map_err(|error| {
+            rustls::TLSError::General(format!("Incorrect TEE quote: {:?}", error))
+        })?;
+
+        // Use [`rustls::WebPKIVerifier`] to check certificate correctness.
+        let verifier = rustls::WebPKIVerifier::new();
+        verifier.verify_server_cert(roots, certs, hostname, ocsp)
```

Comment:
So if this was not here, the standard verifier would not be checked? I thought this verifier is in addition to the built in one. Either way, could you add a comment?